lib/protocol/encryption.go

```diff
@@ -0,0 +1,521 @@
+// Copyright (C) 2019 The Syncthing Authors.
+//
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this file,
+// You can obtain one at https://mozilla.org/MPL/2.0/.
+
+package protocol
+
+import (
+	"context"
+	"crypto/rand"
+	"crypto/sha256"
+	"encoding/base32"
+	"errors"
+	"io"
+	"strings"
+	"time"
+
+	"github.com/gogo/protobuf/proto"
+	"github.com/miscreant/miscreant.go"
+	"golang.org/x/crypto/chacha20poly1305"
+	"golang.org/x/crypto/hkdf"
+	"golang.org/x/crypto/scrypt"
+)
+
+const (
+	nonceSize             = 24   // chacha20poly1305.NonceSizeX
+	tagSize               = 16   // chacha20poly1305.Overhead()
+	keySize               = 32   // fits both chacha20poly1305 and AES-SIV
+	minPaddedSize         = 1024 // smallest block we'll allow
+	blockOverhead         = tagSize + nonceSize
+	maxPathComponent      = 200              // characters
+	encryptedDirExtension = ".syncthing-enc" // for top level dirs
+	miscreantAlgo         = "AES-SIV"
+)
+
+// The encryptedModel sits between the encrypted device and the model. It
+// receives encrypted metadata and requests from the untrusted device, so it
+// must decrypt those and answer requests by encrypting the data.
+type encryptedModel struct {
+	Model
```

Comment:
I wonder if this is not dangerous, and this should have a reference to a model but not embed it.

The moment we add a new method to the model, we'll automatically call through to the underlying model if we forget to implement it here, which probably not what we want?

Reply:
agreed, there is no need to embed here